src/hypothesis/searchstrategy/regex.py

```diff
@@ -0,0 +1,150 @@
+# coding=utf-8
+#
+# This file is part of Hypothesis, which may be found at
+# https://github.com/HypothesisWorks/hypothesis-python
+#
+# Most of this work is copyright (C) 2013-2016 David R. MacIver
+# (david@drmaciver.com), but it contains contributions by others. See
+# CONTRIBUTING.rst for a full list of people who may hold copyright, and
+# consult the git log if you need to determine who owns an individual
+# contribution.
+#
+# This Source Code Form is subject to the terms of the Mozilla Public License,
+# v. 2.0. If a copy of the MPL was not distributed with this file, You can
+# obtain one at http://mozilla.org/MPL/2.0/.
+#
+# END HEADER
+
+from __future__ import division, print_function, absolute_import
+
+import re
+import sys
+import string
+from itertools import chain
+
+import hypothesis.strategies as strats
+from hypothesis.searchstrategy import SearchStrategy
+from hypothesis.internal.compat import hunichr, hrange
+
+
+def strategy_concat(strategies):
+    """given a list of strategies yielding strings return a strategy yielding
+    their concatenation."""
+    return strats.tuples(*strategies).map(lambda x: u"".join(x))
+
+
+class RegexStrategy(SearchStrategy):
+    """Strategy for generating strings matching a regular expression.
+
+    Currently does not support \B, positive and negative lookbehind
+    assertions and conditional matching.
+
+    """
+
+    def __init__(self, pattern):
+        parsed = re.sre_parse.parse(pattern)
+        self.cache = {}
+        self.strategies = [self._handle_state(state) for state in parsed]
+
+    def do_draw(self, data):
+        self.cache = {}
+        return u"".join(data.draw(strat) for strat in self.strategies)
+
+    def _categories(self, category):
+        if category == 'category_digit':
+            return string.digits
+        elif category == 'category_not_digit':
+            return string.ascii_letters + string.punctuation
+        elif category == 'category_space':
+            return string.whitespace
+        elif category == 'category_not_space':
+            return string.printable.strip()
```

Comment:
This depends on `string.printable` happening to have all the whitespace at the end. While that's unlikely to change, it might be better to compute these character classes as the set of 0-127 characters matched by the appropriate regular expression, e.g.:
```
not_digit = re.compile('[^\d]')
category_not_digit = ''.join(filter(not_digit.match, string.printable))
```